[_shard] make ShardedTensor be a Tensor and nn.Parameter

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #79825

Differential Revision: [D37707371](https://our.internmc.facebook.com/intern/diff/D37707371)

wanchaol committed:
[_shard] make ShardedTensor be a Tensor and nn.Parameter

[ghstack-poisoned]

kumpera commented on torch/distributed/_shard/sharded_tensor/api.py:
```diff
@@ -67,7 +65,7 @@ def _register_remote_shards(sharded_tensor_id: int, rrefs: List[rpc.RRef[Shard]]
-class ShardedTensor(object):
+class ShardedTensor(torch.Tensor):
     """
     ShardedTensor is an abstraction to represent Tensors that are sharded
```
We should update the docstring to say it's a subclass of Tensor

kumpera commented on torch/distributed/_shard/sharded_tensor/api.py:
```diff
@@ -446,18 +519,24 @@ def _init_from_local_shards(
+        spec = shard_spec._infer_sharding_spec_from_shards_metadata(
+            global_sharded_tensor_metadata.shards_metadata
+        )
+        sharded_tensor = cls.__new__(cls,
```
Can't we use `cls(...)` instead of `cls.__new__(...)`?

wanchaol replied:
I didn't check if using `cls(...)` will trigger `__init__` or not, ideally we don't want to trigger `__init__` here and that's why I specified explicitly, I will check this as a follow up and remove `__new__` if `cls(...)` directly works for us :)

kumpera reviewed (approved):
LGTM but please check the multigpu jobs results before merging as they will have the ST test results in them.

Commit:
Update on "[_shard] make ShardedTensor be a Tensor and nn.Parameter"




[ghstack-poisoned]

wanchaol commented on torch/distributed/_shard/sharded_tensor/api.py:
```diff
@@ -840,186 +928,12 @@ def local_shards(self) -> List[Shard]:
         """
         return self._local_shards
 
-    def size(self, dim: int = None) -> Union[torch.Size, int]:
```
Yes that's correct!

wanchaol committed:
Update on "[_shard] make ShardedTensor be a Tensor and nn.Parameter"




[ghstack-poisoned]

wanchaol commented:
@wanchaol has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37707371).

wanchaol committed:
Update on "[_shard] make ShardedTensor be a Tensor and nn.Parameter"


Differential Revision: [D37707371](https://our.internmc.facebook.com/intern/diff/D37707371)

[ghstack-poisoned]

wanchaol commented:
@wanchaol has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37707371).

wanchaol committed:
Update on "[_shard] make ShardedTensor be a Tensor and nn.Parameter"


Differential Revision: [D37707371](https://our.internmc.facebook.com/intern/diff/D37707371)

[ghstack-poisoned]

wanchaol committed:
Update on "[_shard] make ShardedTensor be a Tensor and nn.Parameter"


Differential Revision: [D37707371](https://our.internmc.facebook.com/intern/diff/D37707371)

[ghstack-poisoned]

wanchaol committed:
Update on "[_shard] make ShardedTensor be a Tensor and nn.Parameter"


Differential Revision: [D37707371](https://our.internmc.facebook.com/intern/diff/D37707371)

[ghstack-poisoned]

wanchaol committed:
Update on "[_shard] make ShardedTensor be a Tensor and nn.Parameter"


Differential Revision: [D37707371](https://our.internmc.facebook.com/intern/diff/D37707371)

[ghstack-poisoned]

wanchaol committed:
Update on "[_shard] make ShardedTensor be a Tensor and nn.Parameter"


Differential Revision: [D37707371](https://our.internmc.facebook.com/intern/diff/D37707371)

[ghstack-poisoned]

wanchaol committed:
Update on "[_shard] make ShardedTensor be a Tensor and nn.Parameter"


Differential Revision: [D37707371](https://our.internmc.facebook.com/intern/diff/D37707371)

[ghstack-poisoned]

wanchaol committed:
Update on "[_shard] make ShardedTensor be a Tensor and nn.Parameter"


Differential Revision: [D37707371](https://our.internmc.facebook.com/intern/diff/D37707371)

[ghstack-poisoned]

janeyx99 commented:
@pytorchbot revert -m "Sorry, reverting for breaking multigpu tests https://hud.pytorch.org/pytorch/pytorch/commit/9c32439a770845793107a584f17f4da70ccb5df0" -c nosignal

Please add ciflow/periodic label to your reland PR to trigger the multigpu tests.
